File: java/rna-transcription/src/main/java/RnaTranscription.java
import static java.util.Map.entry;

import java.util.Map;
import java.util.stream.Collectors;

class RnaTranscription {
  static final Map<Character, Character> DNA_TO_RNA =
      Map.ofEntries(entry('G', 'C'), entry('C', 'G'), entry('T', 'A'), entry('A', 'U'));

  String transcribe(String dnaStrand) {
    if (dnaStrand.chars().anyMatch(dna -> !DNA_TO_RNA.containsKey((char) dna))) {
      throw new IllegalArgumentException("Invalid input");
    }

    return dnaStrand
        .chars()
        .mapToObj(dna -> DNA_TO_RNA.get((char) dna))
        .map(String::valueOf)
        .collect(Collectors.joining());
  }
}

File: cpp/rna-transcription/rna_transcription.cpp
#include "rna_transcription.h"
#include <unordered_map>

namespace rna_transcription
{
std::unordered_map<char, char> DNA_TO_RNA{
    {'A', 'U'},
    {'C', 'G'},
    {'G', 'C'},
    {'T', 'A'}};

char to_rna(char dna)
{
    return DNA_TO_RNA[dna];
}

std::string to_rna(const std::string &dna_strand)
{
    std::string result;
    for (char dna : dna_strand)
    {
        result.push_back(DNA_TO_RNA[dna]);
    }

    return result;
}
} // namespace rna_transcription
